pkg/compressor/compressor.go

```diff
@@ -0,0 +1,134 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compressor
+
+import (
+	"compress/flate"
+	"compress/gzip"
+	"compress/lzw"
+	"compress/zlib"
+	"io"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// CompressSnapshots takes uncompressed data as input and compress the data according to Compression Policy
+// and write the compressed data into one end of pipe.
+func CompressSnapshots(data io.ReadCloser, compressionPolicy string) io.ReadCloser {
+	pReader, pWriter := io.Pipe()
+
+	var gWriter io.WriteCloser
+	log := logrus.NewEntry(logrus.New())
+	logger := log.WithField("actor", "compressor")
+	logger.Infof("compressing the snapshot using %v Compression Policy", compressionPolicy)
+	startCompressTime := time.Now()
+
+	switch compressionPolicy {
+	case GzipCompressionPolicy:
+		gWriter = gzip.NewWriter(pWriter)
+
+	case FlateCompressionPolicy:
+		//1: for best speed
+		//9: for best compression
+		//-2: for Huffman compression
+		gWriter, _ = flate.NewWriter(pWriter, 9) // [-2,9]
+
+	case LzwCompressionPolicy:
+		gWriter = lzw.NewWriter(pWriter, 0, 8)
+
+	case ZlibCompressionPolicy:
+		gWriter = zlib.NewWriter(pWriter)
+
+	}
+
+	go func() {
+		defer pWriter.Close()
+		defer gWriter.Close()
+		defer data.Close()
+		n, err := io.Copy(gWriter, data)
+		if err != nil {
+			logger.Infof("compression failed: %v", err)
+			return
+		}
+		logger.Infof("compressed data size: %v KB", n/1024)
+	}()
+
+	totalCompressTime := time.Now().Sub(startCompressTime).Seconds()
+	logger.Infof("compresses the snapshot in %v seconds", totalCompressTime)
+	return pReader
+}
+
+// DecompressSnapshots take compressed data and snapshotSuffix as input and
+// it decompresses the data according to compression Policy and return uncompressed data.
+func DecompressSnapshots(data io.ReadCloser, snapshotSuffix string) io.ReadCloser {
+	var unCompressData io.ReadCloser
+	var err error
+
+	log := logrus.NewEntry(logrus.New())
+	logger := log.WithField("actor", "de-compressor")
+	logger.Infof("decompressing the snapshot with %v Policy", snapshotSuffix)
+	startDeCompressTime := time.Now()
+
+	switch snapshotSuffix {
+	case ZlibCompressionExtension:
+		unCompressData, err = zlib.NewReader(data)
+		if err != nil {
+			logger.Infof("Unable to decompress: %v", err)
+			return data
+		}
+
+	case GzipCompressionExtension:
+		unCompressData, err = gzip.NewReader(data)
+		if err != nil {
+			logger.Infof("Unable to decompress: %v", err)
+			return data
+		}
+
+	case LzwCompressionExtension:
+		unCompressData = lzw.NewReader(data, 0, 8) // 0 - LSB and 1- for MSB
+
+	case FlateCompressionExtension:
+		unCompressData = flate.NewReader(data)
+
+	}
+
+	totalDeCompressTime := time.Now().Sub(startDeCompressTime).Seconds()
+	logger.Infof("decompress the snapshot in %v seconds", totalDeCompressTime)
```

Comment:
I agree with your compression comment but here isn't this works fine because in decompression everything is sequential as compared to compression where we run also separate thread.

Reply:
We need to log only the restoration time for each individual file. That way we can compare the times with or without compression. There is no good way to split the time for compression with the time for download or restoration for a single run.